File: NSTU-lab3/lab3_go/containers/hashmap.go
package containers

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"hash/fnv"
	"io"
)

// HashMap - хеш-таблица с двойным хешированием
type HashMap struct {
	table    []entry
	capacity int
	count    int
}

type entry struct {
	key      string
	value    string
	occupied bool
}

const defaultCapacity = 101

// NewHashMap создаёт новую хеш-таблицу
func NewHashMap(initialCapacity int) *HashMap {
	if initialCapacity <= 0 {
		initialCapacity = defaultCapacity
	}
	return &HashMap{
		table:    make([]entry, initialCapacity),
		capacity: initialCapacity,
		count:    0,
	}
}

// hash1 первая хеш-функция
func (h *HashMap) hash1(key string) int {
	hasher := fnv.New32a()
	hasher.Write([]byte(key))
	return int(hasher.Sum32()) % h.capacity
}

// hash2 вторая хеш-функция (для двойного хеширования)
func (h *HashMap) hash2(key string) int {
	hasher := fnv.New32a()
	hasher.Write([]byte(key))
	h2 := int(hasher.Sum32()) % (h.capacity - 1)
	if h2 == 0 {
		return 1
	}
	return h2
}

// findIndex ищет индекс для ключа
func (h *HashMap) findIndex(key string) (int, bool) {
	h1 := h.hash1(key)
	h2 := h.hash2(key)
	index := h1

	for i := 0; i < h.capacity; i++ {
		if !h.table[index].occupied {
			return index, false
		}
		if h.table[index].key == key {
			return index, true
		}
		index = (h1 + (i+1)*h2) % h.capacity
	}
	return -1, false
}

// Put добавляет или обновляет пару ключ-значение
func (h *HashMap) Put(key string, value string) error {
	if h.count >= h.capacity {
		return fmt.Errorf("хеш-таблица заполнена")
	}

	index, found := h.findIndex(key)
	if index == -1 {
		return fmt.Errorf("хеш-таблица заполнена")
	}

	if found {
		h.table[index].value = value
	} else {
		h.table[index] = entry{key: key, value: value, occupied: true}
		h.count++
	}
	return nil
}

// Get возвращает значение по ключу
func (h *HashMap) Get(key string) (string, error) {
	index, found := h.findIndex(key)
	if !found {
		return "", fmt.Errorf("ключ не найден")
	}
	return h.table[index].value, nil
}

// Contains проверяет наличие ключа
func (h *HashMap) Contains(key string) bool {
	_, found := h.findIndex(key)
	return found
}

// Remove удаляет пару по ключу
func (h *HashMap) Remove(key string) bool {
	index, found := h.findIndex(key)
	if !found {
		return false
	}
	h.table[index].occupied = false
	h.count--
	return true
}

// Clear очищает таблицу
func (h *HashMap) Clear() {
	for i := range h.table {
		h.table[i].occupied = false
	}
	h.count = 0
}

// Size возвращает количество элементов
func (h *HashMap) Size() int {
	return h.count
}

// Print выводит хеш-таблицу
func (h *HashMap) Print() string {
	result := "HashMap {\n"
	for i, e := range h.table {
		if e.occupied {
			result += fmt.Sprintf("  [%d] %s => %s\n", i, e.key, e.value)
		}
	}
	result += fmt.Sprintf("} (size: %d)", h.count)
	return result
}

// SaveToBinary сохраняет хеш-таблицу в бинарном формате
func (h *HashMap) SaveToBinary(w io.Writer) error {
	if err := binary.Write(w, binary.LittleEndian, int32(h.count)); err != nil {
		return err
	}

	for _, e := range h.table {
		if e.occupied {
			// Записываем длину ключа и сам ключ
			if err := binary.Write(w, binary.LittleEndian, int32(len(e.key))); err != nil {
				return err
			}
			if _, err := w.Write([]byte(e.key)); err != nil {
				return err
			}

			// Записываем длину значения и само значение
			if err := binary.Write(w, binary.LittleEndian, int32(len(e.value))); err != nil {
				return err
			}
			if _, err := w.Write([]byte(e.value)); err != nil {
				return err
			}
		}
	}
	return nil
}

// LoadFromBinary загружает хеш-таблицу из бинарного формата
func (h *HashMap) LoadFromBinary(r io.Reader) error {
	h.Clear()

	var count int32
	if err := binary.Read(r, binary.LittleEndian, &count); err != nil {
		return err
	}

	for i := 0; i < int(count); i++ {
		// Читаем ключ
		var keyLen int32
		if err := binary.Read(r, binary.LittleEndian, &keyLen); err != nil {
			return err
		}
		keyBytes := make([]byte, keyLen)
		if _, err := io.ReadFull(r, keyBytes); err != nil {
			return err
		}

		// Читаем значение
		var valueLen int32
		if err := binary.Read(r, binary.LittleEndian, &valueLen); err != nil {
			return err
		}
		valueBytes := make([]byte, valueLen)
		if _, err := io.ReadFull(r, valueBytes); err != nil {
			return err
		}

		if err := h.Put(string(keyBytes), string(valueBytes)); err != nil {
			return err
		}
	}
	return nil
}

// ToJSON сериализует HashMap в JSON
func (h *HashMap) ToJSON() (string, error) {
	data := make(map[string]string)
	for i := 0; i < h.capacity; i++ {
		if h.table[i].occupied {
			data[h.table[i].key] = h.table[i].value
		}
	}
	bytes, err := json.Marshal(data)
	if err != nil {
		return "", err
	}
	return string(bytes), nil
}

// FromJSON десериализует HashMap из JSON
func (h *HashMap) FromJSON(jsonStr string) error {
	h.Clear()
	var data map[string]string
	if err := json.Unmarshal([]byte(jsonStr), &data); err != nil {
		return err
	}
	for key, value := range data {
		if err := h.Put(key, value); err != nil {
			return err
		}
	}
	return nil
}

File: NSTU-lab3/lab3/src/containers/hashmap.cpp
#ifndef HASHMAP_CPP
#define HASHMAP_CPP

#include <cstdint>
#include <fstream>
#include "../binary_serialization.h"
#include "hash.h"

template<typename K, typename V>
HashMap<K, V>::Entry::Entry() : occupied(false) {}

template<typename K, typename V>
HashMap<K, V>::Entry::Entry(const K& k, const V& v) : key(k), value(v), occupied(true) {}

template<typename K, typename V>
HashMap<K, V>::HashMap(size_t initialCapacity) 
    : capacity(initialCapacity), count(0) {
    table.resize(capacity);
}

template<typename K, typename V>
size_t HashMap<K, V>::hash1(const K& key) const {
    std::hash<K> hasher;
    return hasher(key) % capacity;
}

template<typename K, typename V>
size_t HashMap<K, V>::hash2(const K& key) const {
    std::hash<K> hasher;
    size_t h = hasher(key) % (capacity - 1);
    return h == 0 ? 1 : h;
}

template<typename K, typename V>
bool HashMap<K, V>::findIndex(const K& key, size_t& index) const {
    size_t h1 = hash1(key);
    size_t h2 = hash2(key);
    index = h1;
    
    for (size_t i = 0; i < capacity; i++) {
        if (!table[index].occupied) {
            return false;
        }
        if (table[index].key == key) {
            return true;
        }
        index = (h1 + (i + 1) * h2) % capacity;
    }
    throw std::runtime_error("Хеш-таблица заполнена");
}

template<typename K, typename V>
void HashMap<K, V>::put(const K& key, const V& value) {
    if (count >= capacity) {
        throw std::runtime_error("Хеш-таблица заполнена");
    }
    size_t index;
    bool found = findIndex(key, index);
    if (found) {
        table[index].value = value;
    } else {
        table[index] = Entry(key, value);
        count++;
    }
}

template<typename K, typename V>
V HashMap<K, V>::get(const K& key) const {
    size_t index;
    if (findIndex(key, index)) {
        return table[index].value;
    }
    throw std::runtime_error("Ключ не найден");
}

template<typename K, typename V>
bool HashMap<K, V>::contains(const K& key) const {
    size_t index;
    return findIndex(key, index);
}

template<typename K, typename V>
bool HashMap<K, V>::remove(const K& key) {
    size_t index;
    if (findIndex(key, index)) {
        table[index].occupied = false;
        count--;
        return true;
    }
    return false;
}

template<typename K, typename V>
void HashMap<K, V>::clear() {
    for (auto& entry : table) {
        entry.occupied = false;
    }
    count = 0;
}

template<typename K, typename V>
size_t HashMap<K, V>::size() const {
    return count;
}

template<typename K, typename V>
bool HashMap<K, V>::empty() const {
    return count == 0;
}

template<typename K, typename V>
void HashMap<K, V>::print(std::ostream& os) const {
    os << "HashMap {\n";
    for (size_t i = 0; i < capacity; i++) {
        if (table[i].occupied) {
            os << "  [" << i << "] " << table[i].key 
               << " => " << table[i].value << "\n";
        }
    }
    os << "} (size: " << count << ")";
}

// Бинарная сериализация для HashMap<string, string>
template<>
inline void HashMap<std::string, std::string>::saveToBinary(std::ofstream& out) const {
    // Записываем количество элементов
    uint32_t sz = static_cast<uint32_t>(count);
    out.write(reinterpret_cast<const char*>(&sz), sizeof(sz));
    
    // Записываем каждую пару ключ-значение
    for (size_t i = 0; i < capacity; i++) {
        if (table[i].occupied) {
            // Записываем ключ
            uint32_t keyLen = static_cast<uint32_t>(table[i].key.length());
            out.write(reinterpret_cast<const char*>(&keyLen), sizeof(keyLen));
            out.write(table[i].key.c_str(), keyLen);
            
            // Записываем значение
            uint32_t valLen = static_cast<uint32_t>(table[i].value.length());
            out.write(reinterpret_cast<const char*>(&valLen), sizeof(valLen));
            out.write(table[i].value.c_str(), valLen);
        }
    }
}

template<>
inline void HashMap<std::string, std::string>::loadFromBinary(std::ifstream& in) {
    clear();
    
    // Читаем количество элементов
    uint32_t sz;
    in.read(reinterpret_cast<char*>(&sz), sizeof(sz));
    
    // Читаем каждую пару
    for (uint32_t i = 0; i < sz; i++) {
        // Читаем ключ
        uint32_t keyLen;
        in.read(reinterpret_cast<char*>(&keyLen), sizeof(keyLen));
        std::string key(keyLen, '\0');
        in.read(&key[0], keyLen);
        
        // Читаем значение
        uint32_t valLen;
        in.read(reinterpret_cast<char*>(&valLen), sizeof(valLen));
        std::string value(valLen, '\0');
        in.read(&value[0], valLen);
        
        put(key, value);
    }
}

// Общий шаблон для других типов
template<typename K, typename V>
void HashMap<K, V>::saveToBinary(std::ofstream& out) const {
    uint32_t sz = static_cast<uint32_t>(count);
    out.write(reinterpret_cast<const char*>(&sz), sizeof(sz));
    
    for (size_t i = 0; i < capacity; i++) {
        if (table[i].occupied) {
            out.write(reinterpret_cast<const char*>(&table[i].key), sizeof(K));
            out.write(reinterpret_cast<const char*>(&table[i].value), sizeof(V));
        }
    }
}

template<typename K, typename V>
void HashMap<K, V>::loadFromBinary(std::ifstream& in) {
    clear();
    uint32_t sz;
    in.read(reinterpret_cast<char*>(&sz), sizeof(sz));
    
    for (uint32_t i = 0; i < sz; i++) {
        K key;
        V value;
        in.read(reinterpret_cast<char*>(&key), sizeof(K));
        in.read(reinterpret_cast<char*>(&value), sizeof(V));
        put(key, value);
    }
}

#endif
